File: IntroToJava/Library.java
// Library.java
// Manages the full catalog of Book and Comic objects.
// Handles file I/O using BufferedReader and BufferedWriter,
// and provides all methods for adding, removing, editing,
// searching, and filtering entries.

import java.io.*;
import java.util.ArrayList;

public class Library {

    // The main data structure: a list of Book objects (may include Comic instances)
    private ArrayList<Book> books;

    // Constructor: initializes the catalog as an empty list
    public Library() {
        this.books = new ArrayList<>();
    }

    // Loads the catalog from a pipe-separated text file.
    // Each line represents one entry; the first field determines the type.
    public void load(String filename) {
        File file = new File(filename);
        if (!file.exists()) {
            System.out.println("  No existing catalog found. Starting fresh.");
            return;
        }
        try (BufferedReader reader = new BufferedReader(new FileReader(file))) {
            String line;
            while ((line = reader.readLine()) != null) {
                // Skip empty lines
                if (line.trim().isEmpty()) continue;
                String[] fields = line.split("\\|", -1);
                String type = fields[0];
                if (type.equals("comic") && fields.length >= 10) {
                    // Reconstruct a Comic object from 10 fields
                    Comic comic = new Comic(
                        fields[1], fields[2], fields[3], fields[4],
                        Double.parseDouble(fields[5]),
                        Boolean.parseBoolean(fields[6]),
                        fields[7], fields[8], fields[9]
                    );
                    books.add(comic);
                } else if (fields.length >= 7) {
                    // Reconstruct a Book object from 7 fields
                    Book book = new Book(
                        fields[1], fields[2], fields[3], fields[4],
                        Double.parseDouble(fields[5]),
                        Boolean.parseBoolean(fields[6])
                    );
                    books.add(book);
                }
            }
            System.out.println("  Catalog loaded: " + books.size() + " entry/entries found.");
        } catch (IOException e) {
            System.out.println("  Error reading file: " + e.getMessage());
        }
    }

    // Writes all entries to the file, one per line in pipe-separated format
    public void save(String filename) {
        try (BufferedWriter writer = new BufferedWriter(new FileWriter(filename))) {
            for (Book book : books) {
                writer.write(book.toFileLine());
                writer.newLine();
            }
        } catch (IOException e) {
            System.out.println("  Error saving file: " + e.getMessage());
        }
    }

    // Adds a new entry to the catalog and saves
    public void addBook(Book book, String filename) {
        books.add(book);
        save(filename);
        System.out.println("  '" + book.getTitle() + "' added to the catalog.");
    }

    // Removes an entry by title (case-insensitive exact match) and saves
    public void removeBook(String title, String filename) {
        for (Book book : books) {
            if (book.getTitle().equalsIgnoreCase(title)) {
                books.remove(book);
                save(filename);
                System.out.println("  '" + book.getTitle() + "' removed from the catalog.");
                return;
            }
        }
        System.out.println("  Entry not found.");
    }

    // Edits a specific field of an entry identified by title.
    // Comic-specific fields are routed to updateIllustrator(),
    // all other fields are routed to update().
    public void editBook(String title, String field, String newValue, String filename) {
        String[] comicFields = {"illustrator", "illustratorGender", "style"};
        for (Book book : books) {
            if (book.getTitle().equalsIgnoreCase(title)) {
                boolean isComicField = false;
                for (String cf : comicFields) {
                    if (cf.equals(field)) { isComicField = true; break; }
                }
                if (isComicField && book instanceof Comic) {
                    ((Comic) book).updateIllustrator(field, newValue);
                } else {
                    book.update(field, newValue);
                }
                save(filename);
                return;
            }
        }
        System.out.println("  Entry not found.");
    }

    // Marks an entry as read by title and saves
    public void markRead(String title, String filename) {
        for (Book book : books) {
            if (book.getTitle().equalsIgnoreCase(title)) {
                book.markAsRead();
                save(filename);
                System.out.println("  '" + book.getTitle() + "' marked as read.");
                return;
            }
        }
        System.out.println("  Entry not found.");
    }

    // Displays all entries in the catalog
    public void displayAll() {
        if (books.isEmpty()) {
            System.out.println("  The catalog is empty.");
            return;
        }
        System.out.println("\n  === Catalog (" + books.size() + " entry/entries) ===");
        for (Book book : books) {
            book.display();
        }
    }

    // Calculates and displays the total purchase value of all entries
    public void totalValue() {
        double total = 0.0;
        for (Book book : books) {
            total += book.getPrice();
        }
        System.out.printf("%n  Total catalog value: %.2f EUR%n", total);
        System.out.println("  (" + books.size() + " entry/entries counted)");
    }

    // Searches entries by title using a case-insensitive partial match
    public void searchByTitle(String query) {
        ArrayList<Book> results = new ArrayList<>();
        for (Book book : books) {
            if (book.getTitle().toLowerCase().contains(query.toLowerCase())) {
                results.add(book);
            }
        }
        printResults(results, "title containing '" + query + "'");
    }

    // Searches entries by author name using a case-insensitive partial match
    public void searchByAuthor(String query) {
        ArrayList<Book> results = new ArrayList<>();
        for (Book book : books) {
            if (book.getAuthor().toLowerCase().contains(query.toLowerCase())) {
                results.add(book);
            }
        }
        printResults(results, "author containing '" + query + "'");
    }

    // Filters entries by type: "novel" or "comic"
    public void filterByType(String type) {
        ArrayList<Book> results = new ArrayList<>();
        for (Book book : books) {
            if (book.getType().equalsIgnoreCase(type)) {
                results.add(book);
            }
        }
        printResults(results, "type '" + type + "'");
    }

    // Filters comic entries by visual style
    public void filterByStyle(String style) {
        ArrayList<Book> results = new ArrayList<>();
        for (Book book : books) {
            if (book instanceof Comic) {
                Comic comic = (Comic) book;
                if (comic.getStyle().equalsIgnoreCase(style)) {
                    results.add(book);
                }
            }
        }
        printResults(results, "style '" + style + "'");
    }

    // Filters all entries by author gender
    public void filterByAuthorGender(String gender) {
        ArrayList<Book> results = new ArrayList<>();
        for (Book book : books) {
            if (book.getAuthorGender().equalsIgnoreCase(gender)) {
                results.add(book);
            }
        }
        printResults(results, "author gender '" + gender + "'");
    }

    // Filters comic entries by illustrator gender
    public void filterByIllustratorGender(String gender) {
        ArrayList<Book> results = new ArrayList<>();
        for (Book book : books) {
            if (book instanceof Comic) {
                Comic comic = (Comic) book;
                if (comic.getIllustratorGender().equalsIgnoreCase(gender)) {
                    results.add(book);
                }
            }
        }
        printResults(results, "illustrator gender '" + gender + "'");
    }

    // Private helper: prints a list of results with a label.
    // The private modifier signals this method is for internal use only.
    private void printResults(ArrayList<Book> results, String label) {
        if (results.isEmpty()) {
            System.out.println("  No results found for " + label + ".");
            return;
        }
        System.out.println("\n  === Results for " + label + " (" + results.size() + " found) ===");
        for (Book book : results) {
            book.display();
        }
    }
}

File: IntroToPython/Library.py
import json
import os
from Book import Book, Comic

class Library:
    """Manages the full collection of Book and Comic objects."""

    def __init__(self):
        self.books = []

    def load(self, filename):
        """Load books from a JSON file if it exists."""
        if os.path.exists(filename):
            with open(filename, "r", encoding="utf-8") as f:
                data = json.load(f)
            for entry in data:
                if entry["type"] == "comic":
                    book = Comic(
                        entry["title"], entry["author"], entry["author_gender"],
                        entry["publisher"], entry["price"], entry["read_status"],
                        entry["illustrator"], entry["illustrator_gender"], entry["style"]
                    )
                else:
                    book = Book(
                        entry["title"], entry["author"], entry["author_gender"],
                        entry["publisher"], entry["price"], entry["read_status"]
                    )
                self.books.append(book)
            print(f"  Library loaded: {len(self.books)} book(s) found.")
        else:
            print("  No existing library found. Starting fresh.")

    def save(self, filename):
        """Save all books to a JSON file."""
        with open(filename, "w", encoding="utf-8") as f:
            json.dump([b.to_dict() for b in self.books], f, indent=2, ensure_ascii=False)

    def add_book(self, book, filename):
        """Add a book to the collection and save."""
        self.books.append(book)
        self.save(filename)
        print(f"  '{book.title}' added to the library.")

    def remove_book(self, title, filename):
        """Remove a book by title and save."""
        for book in self.books:
            if book.title.lower() == title.lower():
                self.books.remove(book)
                self.save(filename)
                print(f"  '{book.title}' removed from the library.")
                return
        print("  Book not found.")

    def edit_book(self, title, field, new_value, filename):
        """Edit a field of a book identified by title."""
        comic_fields = ["illustrator", "illustrator_gender", "style"]
        for book in self.books:
            if book.title.lower() == title.lower():
                if field in comic_fields and isinstance(book, Comic):
                    book.update_illustrator(field, new_value)
                else:
                    book.update(field, new_value)
                self.save(filename)
                return
        print("  Book not found.")

    def mark_read(self, title, filename):
        """Mark a book as read by title."""
        for book in self.books:
            if book.title.lower() == title.lower():
                book.mark_as_read()
                self.save(filename)
                print(f"  '{book.title}' marked as read.")
                return
        print("  Book not found.")

    def display_all(self):
        """Display all books in the library."""
        if not self.books:
            print("  The library is empty.")
            return
        print(f"\n  === Library ({len(self.books)} book(s)) ===")
        for book in self.books:
            book.display()

    def search_by_title(self, query):
        """Search books by title (case-insensitive partial match)."""
        results = [b for b in self.books if query.lower() in b.title.lower()]
        self._print_results(results, f"title containing '{query}'")

    def search_by_author(self, query):
        """Search books by author name (case-insensitive partial match)."""
        results = [b for b in self.books if query.lower() in b.author.lower()]
        self._print_results(results, f"author containing '{query}'")

    def filter_by_type(self, book_type):
        """Filter books by type: 'novel' or 'comic'."""
        results = [b for b in self.books if b.type == book_type.lower()]
        self._print_results(results, f"type '{book_type}'")

    def filter_by_style(self, style):
        """Filter comics by visual style."""
        results = [b for b in self.books
                   if isinstance(b, Comic) and b.style.lower() == style.lower()]
        self._print_results(results, f"style '{style}'")

    def filter_by_author_gender(self, gender):
        """Filter books by author gender."""
        results = [b for b in self.books if b.author_gender.lower() == gender.lower()]
        self._print_results(results, f"author gender '{gender}'")

    def filter_by_illustrator_gender(self, gender):
        """Filter comics by illustrator gender."""
        results = [b for b in self.books
                   if isinstance(b, Comic) and b.illustrator_gender.lower() == gender.lower()]
        self._print_results(results, f"illustrator gender '{gender}'")

    def total_value(self):
        """Calculate and display the total value of the library."""
        total = sum(b.price for b in self.books)
        print(f"\n  Total library value: {total:.2f} EUR")
        print(f"  ({len(self.books)} book(s) counted)")

    def _print_results(self, results, label):
        """Helper method to print a list of results."""
        if not results:
            print(f"  No results found for {label}.")
            return
        print(f"\n  === Results for {label} ({len(results)} found) ===")
        for book in results:
            book.display()